js/core/template_manager.js

```diff
@@ -0,0 +1,277 @@
+    }
+
+    dispose() {
+        this._tempTemplates.forEach(tempTemplate => {
+            tempTemplate.template.dispose && tempTemplate.template.dispose();
+        });
+        this._tempTemplates = [];
+    }
+
+    extractTemplates($el) {
+        const templates = this._extractTemplates($el);
+        const anonymousTemplateMeta = this._extractAnonymousTemplate($el);
+        return { templates, anonymousTemplateMeta };
+    }
+
+    _extractTemplates($el) {
+        const templateElements = $el.contents().filter(TEMPLATE_SELECTOR);
+        const templatesMap = {};
+
+        templateElements.each((_, template) => {
+            const templateOptions = getElementOptions(template).dxTemplate;
+
+            if(!templateOptions) {
+                return;
+            }
+
+            if(!templateOptions.name) {
+                throw Error('E0023');
+            }
+
+            $(template).addClass(TEMPLATE_WRAPPER_CLASS).detach();
+            templatesMap[templateOptions.name] = templatesMap[templateOptions.name] || [];
+            templatesMap[templateOptions.name].push(template);
+        });
+
+        const templates = [];
+        for(let templateName in templatesMap) {
+            const deviceTemplate = TemplateManager._findTemplateByDevice(templatesMap[templateName]);
+            if(deviceTemplate) {
+                templates.push({
+                    name: templateName,
+                    template: this._createTemplate(deviceTemplate),
+                });
+            }
+        }
+        return templates;
+    }
+
+    _extractAnonymousTemplate($el) {
+        const $anonymousTemplate = $el.contents().detach();
+
+        const $notJunkTemplateContent = $anonymousTemplate.filter((_, element) => {
+            const isTextNode = element.nodeType === TEXT_NODE;
+            const isEmptyText = $(element).text().trim().length < 1;
+
+            return !(isTextNode && isEmptyText);
+        });
+        const onlyJunkTemplateContent = $notJunkTemplateContent.length < 1;
+
+        return !onlyJunkTemplateContent
+            ? { template: this._createTemplate($anonymousTemplate), name: this._anonymousTemplateName }
+            : {};
+    }
+
+    _createTemplateIfNeeded(templateSource) {
+        const cachedTemplate = this._tempTemplates.filter(tempTemplate =>
+            tempTemplate.source === TemplateManager.templateKey(templateSource)
+        )[0];
+        if(cachedTemplate) return cachedTemplate.template;
+
+        const template = this._createTemplate(templateSource);
+        this._tempTemplates.push({ template, source: TemplateManager.templateKey(templateSource) });
+        return template;
+    }
+
+    _createTemplate(templateSource) {
+        return this._createElement(TemplateManager.validateTemplateSource(templateSource));
+    }
+
+    getTemplate(templateSource, templates, { isAsyncTemplate, skipTemplates }, context) {
+        if(isFunction(templateSource)) {
```

Comment:
Invert condition to decrease main block indetation
```suggestion
        if(!isFunction(templateSource)) {
          return this._acquireTemplate(
            templateSource,
            this._createTemplateIfNeeded.bind(this),
            templates,
            isAsyncTemplate,
            skipTemplates
            );
          }
          
          return new FunctionTemplate((options) => {
          
          ...
```